<a href="https://colab.research.google.com/github/jdy5294/Bigdata_Analyst/blob/main/4_2_%EB%AA%A8%EC%9D%98%EA%B3%A0%EC%82%AC_%EC%9E%91%EC%97%85%ED%98%95_2%ED%9A%8C.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### 1. 제1유형

데이터셋: HR-Employee-Attrition.csv

      - Attrition: Yes-퇴사, No-퇴사하지 않음

#### 1-1.

데이터에서 'Attrition'은 퇴사여부에 대한 종속변수에 해당한다. 종속변수의 값을 다음과 같이 수치로 변환해 새 컬럼으로 추가하고 각 범주별 레코드 수를 계산하시오. (Yes → 1, No → 0)

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [11]:
import numpy as np
import pandas as pd

df = pd.read_csv("/content/drive/MyDrive/dataset/python_csv/HR-Employee-Attrition.csv")
df

Age Attrition     BusinessTravel  DailyRate              Department  \
0      41       Yes      Travel_Rarely       1102                   Sales   
1      49        No  Travel_Frequently        279  Research & Development   
2      37       Yes      Travel_Rarely       1373  Research & Development   
3      33        No  Travel_Frequently       1392  Research & Development   
4      27        No      Travel_Rarely        591  Research & Development   
...   ...       ...                ...        ...                     ...   
1465   36        No  Travel_Frequently        884  Research & Development   
1466   39        No      Travel_Rarely        613  Research & Development   
1467   27        No      Travel_Rarely        155  Research & Development   
1468   49        No  Travel_Frequently       1023                   Sales   
1469   34        No      Travel_Rarely        628  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  \
0                    1          2  Life Sciences              1   
1                    8          1  Life Sciences              1   
2                    2          2          Other              1   
3                    3          4  Life Sciences              1   
4                    2          1        Medical              1   
...                ...        ...            ...            ...   
1465                23          2        Medical              1   
1466                 6          1        Medical              1   
1467                 4          3  Life Sciences              1   
1468                 2          3        Medical              1   
1469                 8          3        Medical              1   

      EmployeeNumber  ...  RelationshipSatisfaction StandardHours  \
0                  1  ...                         1            80   
1                  2  ...                         4            80   
2                  4  ...                         2            80   
3                  5  ...                         3            80   
4                  7  ...                         4            80   
...              ...  ...                       ...           ...   
1465            2061  ...                         3            80   
1466            2062  ...                         1            80   
1467            2064  ...                         2            80   
1468            2065  ...                         4            80   
1469            2068  ...                         1            80   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                    0                  8                      0   
1                    1                 10                      3   
2                    0                  7                      3   
3                    0                  8                      3   
4                    1                  6                      3   
...                ...                ...                    ...   
1465                 1                 17                      3   
1466                 1                  9                      5   
1467                 1                  6                      0   
1468                 0                 17                      3   
1469                 0                  6                      3   

     WorkLifeBalance  YearsAtCompany YearsInCurrentRole  \
0                  1               6                  4   
1                  3              10                  7   
2                  3               0                  0   
3                  3               8                  7   
4                  3               2                  2   
...              ...             ...                ...   
1465               3               5                  2   
1466               3               7                  7   
1467               3               6                  2   
1468               2               9    

In [12]:
from sklearn.preprocessing import LabelEncoder

df['AttritionNew'] = LabelEncoder().fit_transform(df['Attrition'])
df['AttritionNew'].value_counts()

0    1233
1     237
Name: AttritionNew, dtype: int64

#### 1-2.

데이터셋의 데이터타입별 컬럼 개수를 계산하고, 범주형 변수 중 유일한 값을 1개만 가지고 있는 컬럼(변수)를 찾아내어 그 변수를 데이터에서 제거하시오.

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 36 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

In [26]:
df_object = df.select_dtypes('object')
df_object.nunique().sort_values()

Over18            1
Attrition         2
Gender            2
OverTime          2
BusinessTravel    3
Department        3
MaritalStatus     3
EducationField    6
JobRole           9
dtype: int64

In [28]:
df_removed = df.drop('Over18', axis=1)
df_removed.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

#### 1-3.

원래 데이터에서 숫자형인 컬럼만 추출하여 새로운 데이터프레임을 생성하고, 각 변수간의 상관계수를 구하고 상관계수가 0.9 이상인 두 개의 컬럼을 찾아내어 그 변수 중 하나를 제거하시오.

In [30]:
import numpy as np
import pandas as pd

df = pd.read_csv("/content/drive/MyDrive/dataset/python_csv/HR-Employee-Attrition.csv")
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

In [39]:
df_New = df.select_dtypes('int')
df_New

Age  DailyRate  DistanceFromHome  Education  EmployeeCount  \
0      41       1102                 1          2              1   
1      49        279                 8          1              1   
2      37       1373                 2          2              1   
3      33       1392                 3          4              1   
4      27        591                 2          1              1   
...   ...        ...               ...        ...            ...   
1465   36        884                23          2              1   
1466   39        613                 6          1              1   
1467   27        155                 4          3              1   
1468   49       1023                 2          3              1   
1469   34        628                 8          3              1   

      EmployeeNumber  EnvironmentSatisfaction  HourlyRate  JobInvolvement  \
0                  1                        2          94               3   
1                  2                        3          61               2   
2                  4                        4          92               2   
3                  5                        4          56               3   
4                  7                        1          40               3   
...              ...                      ...         ...             ...   
1465            2061                        3          41               4   
1466            2062                        4          42               2   
1467            2064                        2          87               4   
1468            2065                        4          63               2   
1469            2068                        2          82               4   

      JobLevel  ...  RelationshipSatisfaction  StandardHours  \
0            2  ...                         1             80   
1            2  ...                         4             80   
2            1  ...                         2             80   
3            1  ...                         3             80   
4            1  ...                         4             80   
...        ...  ...                       ...            ...   
1465         2  ...                         3             80   
1466         3  ...                         1             80   
1467         2  ...                         2             80   
1468         2  ...                         4             80   
1469         2  ...                         1             80   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                    0                  8                      0   
1                    1                 10                      3   
2                    0                  7                      3   
3                    0                  8                      3   
4                    1                  6                      3   
...                ...                ...                    ...   
1465                 1                 17                      3   
1466                 1                  9                      5   
1467                 1                  6                      0   
1468                 0                 17                      3   
1469                 0                  6                      3   

      WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                   1               6                   4   
1                   3              10                   7   
2                   3               0                   0   
3                   3               8                   7   
4                   3               2                   2   
...               ...             ...                 ...   
1465                3               5                   2   
1466                3               7                   7   
1467                3               6                   2   
1468                2               9                   6   
1469   

In [65]:
from itertools import combinations
from scipy.stats import pearsonr

num_feat = df_New.columns.values
comb_num_feat = np.array(list(combinations(num_feat, 2)))
corr_num_feat = np.array([])

for comb in comb_num_feat:
  #corr = np.corrcoef(df_New[comb[0]], df_New[comb[1]])[0][1]
  corr = pearsonr(df_New[comb[0]], df_New[comb[1]])[0]
  corr_num_feat = np.append(corr_num_feat, corr)

high_corr_num = comb_num_feat[corr_num_feat >= 0.9]
high_corr_num

array([['JobLevel', 'MonthlyIncome']], dtype='<U24')

In [68]:
df_New.drop(high_corr_num[0][0], axis=1, inplace=True)
df_New

Age  DailyRate  DistanceFromHome  Education  EmployeeCount  \
0      41       1102                 1          2              1   
1      49        279                 8          1              1   
2      37       1373                 2          2              1   
3      33       1392                 3          4              1   
4      27        591                 2          1              1   
...   ...        ...               ...        ...            ...   
1465   36        884                23          2              1   
1466   39        613                 6          1              1   
1467   27        155                 4          3              1   
1468   49       1023                 2          3              1   
1469   34        628                 8          3              1   

      EmployeeNumber  EnvironmentSatisfaction  HourlyRate  JobInvolvement  \
0                  1                        2          94               3   
1                  2                        3          61               2   
2                  4                        4          92               2   
3                  5                        4          56               3   
4                  7                        1          40               3   
...              ...                      ...         ...             ...   
1465            2061                        3          41               4   
1466            2062                        4          42               2   
1467            2064                        2          87               4   
1468            2065                        4          63               2   
1469            2068                        2          82               4   

      JobSatisfaction  ...  RelationshipSatisfaction  StandardHours  \
0                   4  ...                         1             80   
1                   2  ...                         4             80   
2                   3  ...                         2             80   
3                   3  ...                         3             80   
4                   2  ...                         4             80   
...               ...  ...                       ...            ...   
1465                4  ...                         3             80   
1466                1  ...                         1             80   
1467                2  ...                         2             80   
1468                2  ...                         4             80   
1469                3  ...                         1             80   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                    0                  8                      0   
1                    1                 10                      3   
2                    0                  7                      3   
3                    0                  8                      3   
4                    1                  6                      3   
...                ...                ...                    ...   
1465                 1                 17                      3   
1466                 1                  9                      5   
1467                 1                  6                      0   
1468                 0                 17                      3   
1469                 0                  6                      3   

      WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                   1               6                   4   
1                   3              10                   7   
2                   3               0                   0   
3                   3               8                   7   
4                   3               2                   2   
...               ...             ...                 ...   
1465                3               5                   2   
1466                3               7                   7   
1467                3               6        

### 2. 제2유형

데이터셋을 이용하여 파킨슨병을 예측하는 모델을 로지스틱 회귀모형을 적용하여 생성하고, 이때 파킨슨병을 예측하는데 영향을 미치는 변수를 중요한 순서대로 3개 선정하시오.

이 모델에서 파킨슨병으로 진단하는 기준(thershold)을 0.5로 했을 때와 0.8로 했을 때의 F1-스코어를 비교하고 해석하시오.

- 분석조건
      - 'name' 컬럼 제거
      - 로지스틱 회귀를 위해 상수항 추가
      - 트레이닝셋과 테스트셋의 비율은 9:1
      - 모델의 최적화 방법론으로 "bfgs" 사용
      - 데이터 정규화는 min-max 스케일러 사용
      - Status는 카테고리 타입으로 변환

#### 2-1. 데이터 불러오기

In [114]:
import numpy as np
import pandas as pd
import statsmodels.api as sm

from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score
from sklearn.linear_model import LogisticRegression

df = pd.read_csv("/content/drive/MyDrive/dataset/python_csv/Parkinsons.csv")
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Data columns (total 24 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   name              195 non-null    object 
 1   MDVP:Fo(Hz)       195 non-null    float64
 2   MDVP:Fhi(Hz)      195 non-null    float64
 3   MDVP:Flo(Hz)      195 non-null    float64
 4   MDVP:Jitter(%)    195 non-null    float64
 5   MDVP:Jitter(Abs)  195 non-null    float64
 6   MDVP:RAP          195 non-null    float64
 7   MDVP:PPQ          195 non-null    float64
 8   Jitter:DDP        195 non-null    float64
 9   MDVP:Shimmer      195 non-null    float64
 10  MDVP:Shimmer(dB)  195 non-null    float64
 11  Shimmer:APQ3      195 non-null    float64
 12  Shimmer:APQ5      195 non-null    float64
 13  MDVP:APQ          195 non-null    float64
 14  Shimmer:DDA       195 non-null    float64
 15  NHR               195 non-null    float64
 16  HNR               195 non-null    float64
 1

#### 2-2. 데이터 전처리

In [112]:
# name 변수 제거
df_preprocessing = df.drop('name', axis = 1, inplace=False)

# Normalization
scaler = MinMaxScaler()
df_processed = pd.DataFrame(scaler.fit_transform(df_preprocessing))

# 상수항 추가
df_processed = sm.add_constant(df_processed, has_constant='add')

# 열 이름 복구
df_processed.columns = np.append('const', df_preprocessing.columns.values)

df_processed

const  MDVP:Fo(Hz)  MDVP:Fhi(Hz)  MDVP:Flo(Hz)  MDVP:Jitter(%)  \
0      1.0     0.184308      0.112592      0.054815        0.195680   
1      1.0     0.198327      0.094930      0.278323        0.254130   
2      1.0     0.165039      0.059128      0.265288        0.280178   
3      1.0     0.165004      0.072927      0.264200        0.263342   
4      1.0     0.161150      0.080909      0.260107        0.354511   
..     ...          ...           ...           ...             ...   
190    1.0     0.499820      0.262986      0.165722        0.092440   
191    1.0     0.705488      0.307974      0.138243        0.125794   
192    1.0     0.502730      0.281413      0.050727        0.378653   
193    1.0     0.642893      0.601807      0.054279        0.181703   
194    1.0     0.733274      0.322794      0.071948        0.126747   

     MDVP:Jitter(Abs)  MDVP:RAP  MDVP:PPQ  Jitter:DDP  MDVP:Shimmer  ...  \
0            0.249012  0.145472  0.247588    0.145288      0.312215  ...   
1            0.288538  0.191233  0.323687    0.191042      0.472887  ...   
2            0.328063  0.229287  0.369239    0.229411      0.390634  ...   
3            0.328063  0.209056  0.324759    0.208862      0.414278  ...   
4            0.407115  0.282755  0.437299    0.282870      0.499452  ...   
..                ...       ...       ...         ...           ...  ...   
190          0.090909  0.093931  0.089496    0.094076      0.286014  ...   
191          0.090909  0.126686  0.107181    0.126826      0.164050  ...   
192          0.288538  0.267823  0.252947    0.267940      0.123608  ...   
193          0.130435  0.145472  0.159700    0.145288      0.122512  ...   
194          0.090909  0.109345  0.120579    0.109327      0.084900  ...   

     Shimmer:DDA       NHR       HNR  status      RPDE       DFA   spread1  \
0       0.332584  0.068307  0.511745     1.0  0.369155  0.960148  0.569875   
1       0.516048  0.059331  0.432577     1.0  0.470830  0.977024  0.703277   
2       0.443317  0.039596  0.496220     1.0  0.404416  1.000000  0.636745   
3       0.475478  0.040997  0.495936     1.0  0.416255  0.975885  0.695627   
4       0.584542  0.054174  0.455499     1.0  0.375159  0.992813  0.762472   
..           ...       ...       ...     ...       ...       ...       ...   
190     0.362306  0.085909  0.450134     0.0  0.447684  0.333127  0.257894   
191     0.221338  0.055543  0.435097     0.0  0.408567  0.434101  0.319956   
192     0.156631  0.338988  0.383728     0.0  0.352318  0.324299  0.212945   
193     0.155989  0.227838  0.429936     0.0  0.454176  0.277579  0.220650   
194     0.110027  0.137919  0.518898     0.0  0.481200  0.358856  0.405161   

      spread2        D2       PPE  
0    0.585765  0.390661  0.497310  
1    0.741337  0.473145  0.671326  
2    0.686371  0.408819  0.596682  
3    0.738089  0.436977  0.671949  
4    0.513798  0.404336  0.757611  
..        ...       ...       ...  
190  0.260408  0.549049  0.183318  
191  0.276956  0.605474  0.257558  
192  0.342577  0.558967  0.180580  
193  0.452885  0.318222  0.163137  
194  0.415095  0.503673  0.215460  

[195 rows x 24 columns]

#### 2-3. 데이터 분석 수행

In [113]:
X = df_processed.drop('status', axis=1, inplace=False)
y = df_processed['status'].astype('category')

X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, test_size=0.1, random_state=2022)
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(175, 23) (20, 23) (175,) (20,)


In [119]:
model = sm.Logit(y_train, X_train)
results = model.fit(method='bfgs', maxiter=1000)
results.summary()

Optimization terminated successfully.
         Current function value: 0.225539
         Iterations: 342
         Function evaluations: 343
         Gradient evaluations: 343


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                 status   No. Observations:                  175
Model:                          Logit   Df Residuals:                      152
Method:                           MLE   Df Model:                           22
Date:                Fri, 23 Jun 2023   Pseudo R-squ.:                  0.5955
Time:                        08:22:00   Log-Likelihood:                -39.469
converged:                       True   LL-Null:                       -97.576
Covariance Type:            nonrobust   LLR p-value:                 8.459e-15
====================================================================================
                       coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------------
const               -5.3533      5.969     -0.897      0.370     -17.053       6.347
MDVP:Fo(Hz)          0.2048      4.018      0.051      0.959      -7.671       8.080
MDVP:Fhi(Hz)        -0.3496      2.313     -0.151      0.880      -4.883       4.184
MDVP:Flo(Hz)        -0.0534      2.298     -0.023      0.981      -4.558       4.451
MDVP:Jitter(%)     -45.7419     42.058     -1.088      0.277    -128.174      36.690
MDVP:Jitter(Abs)   -13.1383     23.796     -0.552      0.581     -59.778      33.502
MDVP:RAP            33.0716   3075.650      0.011      0.991   -5995.091    6061.235
MDVP:PPQ           -33.3312     38.199     -0.873      0.383    -108.199      41.537
Jitter:DDP          42.6551   3077.160      0.014      0.989   -5988.467    6073.777
MDVP:Shimmer       102.5334    123.986      0.827      0.408    -140.475     345.542
MDVP:Shimmer(dB)    12.9021     45.556      0.283      0.777     -76.387     102.191
Shimmer:APQ3       -33.0753   6510.455     -0.005      0.996   -1.28e+04    1.27e+04
Shimmer:APQ5       -38.5700     36.939     -1.044      0.296    -110.968      33.828
MDVP:APQ            24.8434     54.904      0.452      0.651     -82.767     132.453
Shimmer:DDA        -32.8496   6511.661     -0.005      0.996   -1.28e+04    1.27e+04
NHR                  2.6577     23.683      0.112      0.911     -43.761      49.076
HNR                 -0.5534      5.823     -0.095      0.924     -11.966      10.860
RPDE                -3.6073      2.573     -1.402      0.161      -8.650       1.435
DFA                  1.7605      2.325      0.757      0.449      -2.797       6.318
spread1              3.2903     10.293      0.320      0.749     -16.884      23.464
spread2              5.1776      3.079      1.681      0.093      -0.858      11.213
D2                   0.1640      3.515      0.047      0.963      -6.726       7.054
PPE                 18.0515     13.603      1.327      0.185      -8.610      44.713
====================================================================================

Possibly complete quasi-separation: A fraction 0.13 of observations can be
perfectly predicted. This might indicate that there is complete
quasi-separation. In this case some parameters will not be identified.
"""

      p-value가 제일 작은 3개 {'spread2':0.093, 'RPDE':0.161, 'PPE':0.185}

#### 2-4. 임계치에 따른 모델 정확도 비교

In [132]:
pred = results.predict(X_test)
pred80 = (pred>0.8).astype('int')
pred50 = (pred>0.5).astype('int')

print('임계치 0.8의 F1-스코어', f1_score(y_test, pred80))
print('임계치 0.5의 F1-스코어', f1_score(y_test, pred50))

임계치 0.8의 F1-스코어 0.8
임계치 0.5의 F1-스코어 0.846153846153846
